platform/linux-generic/pktio/socket_common.c

```diff
@@ -160,3 +162,131 @@ int link_status_fd(int fd, const char *name)
 
 	return !!(ifr.ifr_flags & IFF_RUNNING);
 }
+
+int link_info_fd(int fd, const char *name, odp_pktio_link_info_t *info)
+{
+	struct ethtool_link_settings hcmd = {.cmd = ETHTOOL_GLINKSETTINGS};
+	struct ethtool_link_settings *ecmd;
+	struct ethtool_pauseparam pcmd = {.cmd = ETHTOOL_GPAUSEPARAM};
+	struct ifreq ifr;
+	int status;
+
+	status = link_status_fd(fd, name);
+	if (status < 0)
+		return -1;
+
+	snprintf(ifr.ifr_name, IF_NAMESIZE, "%s", name);
+
+	/* Link pause status */
+	ifr.ifr_data = (void *)&pcmd;
+	if (ioctl(fd, SIOCETHTOOL, &ifr) && errno != EOPNOTSUPP) {
+		__odp_errno = errno;
+		ODP_ERR("ioctl(SIOCETHTOOL): %s: \"%s\".\n", strerror(errno),
+			ifr.ifr_name);
+		return -1;
+	}
+
+	/* Try to perform handshake and fall back to old API if failed */
+	ifr.ifr_data = (void *)&hcmd;
+	if (ioctl(fd, SIOCETHTOOL, &ifr) < 0) {
+		struct ethtool_cmd ecmd_old = {.cmd = ETHTOOL_GSET};
+
+		ifr.ifr_data = (void *)&ecmd_old;
+		if (ioctl(fd, SIOCETHTOOL, &ifr) < 0) {
+			__odp_errno = errno;
+			ODP_ERR("ioctl(SIOCETHTOOL): %s: \"%s\".\n", strerror(errno), ifr.ifr_name);
+			return -1;
+		}
+
+		memset(info, 0, sizeof(odp_pktio_link_info_t));
+		info->speed = ethtool_cmd_speed(&ecmd_old);
+		if (info->speed == (uint32_t)SPEED_UNKNOWN)
+			info->speed = ODP_PKTIO_LINK_SPEED_UNKNOWN;
+
+		if (ecmd_old.autoneg == AUTONEG_ENABLE)
+			info->autoneg = ODP_PKTIO_LINK_AUTONEG_ON;
+		else if (ecmd_old.autoneg == AUTONEG_DISABLE)
+			info->autoneg = ODP_PKTIO_LINK_AUTONEG_OFF;
+		else
+			info->autoneg = ODP_PKTIO_LINK_AUTONEG_UNKNOWN;
+
+		if (ecmd_old.duplex == DUPLEX_HALF)
+			info->duplex = ODP_PKTIO_LINK_DUPLEX_HALF;
+		else if (ecmd_old.duplex == DUPLEX_FULL)
+			info->duplex = ODP_PKTIO_LINK_DUPLEX_FULL;
+		else
+			info->duplex = ODP_PKTIO_LINK_DUPLEX_UNKNOWN;
+
+		info->pause_rx = pcmd.rx_pause ? ODP_PKTIO_LINK_PAUSE_ON : ODP_PKTIO_LINK_PAUSE_OFF;
+		info->pause_tx = pcmd.tx_pause ? ODP_PKTIO_LINK_PAUSE_ON : ODP_PKTIO_LINK_PAUSE_OFF;
+
+		if (ecmd_old.port == PORT_TP)
+			info->media = "copper";
+		else if (ecmd_old.port == PORT_FIBRE)
+			info->media = "fiber";
+		else if (ecmd_old.port == PORT_OTHER)
+			info->media = "other";
+		else
+			info->media = "unknown";
+
+		info->status = status;
+
+		return 0;
+	}
+
+	if (hcmd.link_mode_masks_nwords >= 0 || hcmd.cmd != ETHTOOL_GLINKSETTINGS) {
+		ODP_ERR("ETHTOOL_GLINKSETTINGS handshake failed\n");
+		return -1;
+	}
+	hcmd.link_mode_masks_nwords = -hcmd.link_mode_masks_nwords;
+
+	/* Read link information */
+	uint8_t ODP_ALIGNED_CACHE data[offsetof(struct ethtool_link_settings, link_mode_masks) +
+				       (3 * sizeof(uint32_t) * hcmd.link_mode_masks_nwords)];
```

Comment:
Some comment about max data size calculation would be helpful. Maybe few sentences from ethtool documentation.